File: examples/kotlin-react/src/components/TodoList.kt
package components

import kotlinx.html.js.onDoubleClickFunction
import model.Todo
import model.TodoFilter
import react.*
import react.dom.li
import react.dom.ul

class TodoList : RComponent<TodoList.Props, TodoList.State>() {

    override fun componentWillMount() {
        setState {
            editingIdx = -1
        }
    }

    override fun RBuilder.render() {
        console.log("TodoList render")

        ul(classes = "todo-list") {
            val filter = props.filter

            props.todos.filter { todo ->
                filter.filter(todo)

            }.forEachIndexed { idx, todo ->
                val isEditing = idx == state.editingIdx

                val classes = when {
                    todo.completed -> "completed"
                    isEditing -> "editing"
                    else -> ""
                }


                li(classes = classes) {
                    attrs.onDoubleClickFunction = {
                        setState {
                            editingIdx = idx
                        }
                    }

                    todoItem(
                            todo = todo,
                            editing = isEditing,
                            endEditing = ::endEditing,
                            removeTodo = { props.removeTodo(todo) },
                            updateTodo = { title, completed ->
                                props.updateTodo(todo.copy(title = title, completed = completed))
                            }
                    )
                }
            }
        }
    }

    private fun endEditing() {
        setState {
            editingIdx = -1
        }
    }

    class Props(var removeTodo: (Todo) -> Unit, var updateTodo: (Todo) -> Unit, var todos: List<Todo>, var filter: TodoFilter) : RProps
    class State(var editingIdx: Int) : RState
}

fun RBuilder.todoList(removeTodo: (Todo) -> Unit, updateTodo: (Todo) -> Unit, todos: List<Todo>, filter: TodoFilter) = child(TodoList::class) {
    attrs.todos = todos
    attrs.removeTodo = removeTodo
    attrs.updateTodo = updateTodo
    attrs.filter = filter
}

File: examples/react-hooks/js/components/main/todo-list.js
function TodoList(props) {
	'use strict';

	const [editing, setEditing] = React.useState(null);
	const [filter] = useFilter();
	const [todos] = useTodos();
	let visibleTodos = todos;

	if (filter === 'active') {
		visibleTodos = todos.filter(({ completed }) => completed === false);
	}

	if (filter === 'completed') {
		visibleTodos = todos.filter(({ completed }) => completed);
	}

	visibleTodos = visibleTodos.map(({ id, text, completed }) => {
		const isEditing = id === editing;

		return (
			<Todo
				key={id}
				id={id}
				text={text}
				isEditing={isEditing}
				setEditing={setEditing}
				completed={completed}
			/>
		);
	});

	return <ul className="todo-list">{visibleTodos}</ul>;
}
